From 13b76a9c8adcbc4e8a9610615a4bac7b10949e85 Mon Sep 17 00:00:00 2001
From: Moritz Angermann <moritz.angermann@gmail.com>
Date: Fri, 7 Jan 2022 17:21:29 +0800
Subject: [PATCH] Fix gitRevFromGit with newer process libraries

This code was initially lifted likely from cardano-node, and now lives in cardano-config, but is replicated here...
It was fixed in cardano config recently by @newhoggy in cardano-config#4. Now we need to fix it here as well.
---
 lib/core/src/Cardano/Wallet/Version/TH.hs | 32 ++++++++++++++---------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/lib/core/src/Cardano/Wallet/Version/TH.hs b/lib/core/src/Cardano/Wallet/Version/TH.hs
index 7d6607a721..516fa5bfa0 100644
--- a/lib/core/src/Cardano/Wallet/Version/TH.hs
+++ b/lib/core/src/Cardano/Wallet/Version/TH.hs
@@ -15,8 +15,10 @@ import Language.Haskell.TH
     ( Exp (..), Lit (..), Q, runIO )
 import System.Exit
     ( ExitCode (..) )
+import System.IO
+    ( hPutStrLn, stderr )
 import System.IO.Error
-    ( ioeGetErrorType, isDoesNotExistErrorType )
+    ( isDoesNotExistError )
 import UnliftIO.Exception
     ( handleJust )
 import UnliftIO.Process
@@ -26,15 +28,19 @@ import UnliftIO.Process
 -- executed, then this will be an empty string.
 gitRevFromGit :: Q Exp
 gitRevFromGit = LitE . StringL <$> runIO runGitRevParse
-  where
-    runGitRevParse :: IO String
-    runGitRevParse = handleJust missingGit (const $ pure "") $ do
-        (exitCode, output, _) <-
-            readProcessWithExitCode "git" ["rev-parse", "--verify", "HEAD"] ""
-        pure $ case exitCode of
-            ExitSuccess -> output
-            _           -> ""
-    missingGit e =
-        if isDoesNotExistErrorType (ioeGetErrorType e)
-        then Just ()
-        else Nothing
+    where
+        runGitRevParse :: IO String
+        runGitRevParse = do
+            (exitCode, output, errorMessage) <- readProcessWithExitCode_ "git" ["rev-parse", "--verify", "HEAD"] ""
+            case exitCode of
+                ExitSuccess -> pure output
+                ExitFailure _ -> do
+                    hPutStrLn stderr $ "WARNING: " ++ errorMessage
+                    pure ""
+
+        readProcessWithExitCode_ :: FilePath -> [String] -> String -> IO (ExitCode, String, String)
+        readProcessWithExitCode_ cmd args input =
+            catch (readProcessWithExitCode cmd args input) $ \e ->
+                if isDoesNotExistError e
+                    then return (ExitFailure 127, "", show e)
+                    else return (ExitFailure 999, "", show e)
